src/python/pants/util/objects.py

```diff
@@ -229,41 +302,31 @@ def create(cls, value=None):
 
     def __new__(cls, *args, **kwargs):
       this_object = super(ChoiceDatatype, cls).__new__(cls, *args, **kwargs)
-
-      field_value = getattr(this_object, field_name)
-
-      cls._check_value(field_value)
-
+      cls._check_value(cls._get_value(this_object))
       return this_object
 
-  return ChoiceDatatype
+    def resolve_for_enum_variant(self, mapping):
```

Comment:
So, rather than "invoking a method", could this just select the relevant dictionary entry, and then the invoking could happen outside? That would make it work for attributes as well as methods.

Reply:
This is much much much better and easier to understand.

Reply:
Done and seems to work! The mention of attributes and methods is very interesting.